File: ruby/combinations.rb
# @param {Integer} n
# @param {Integer} k
# @return {Integer[][]}
def combine(n, k)
    result = []
    combine_helper(n, k, [], result)
    
    result
end

def combine_helper(max, solution_size, solution, result)
    if solution.size == solution_size
        result << solution.clone
    else
        current = solution.last || 0
        (current + 1..max).each do |num|
            solution << num
            combine_helper(max, solution_size, solution, result)
            solution.pop
        end
    end
end

File: javascript/combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    var result = [];
    combineHelper(n, k, [], result);
    
    return result;
};

function combineHelper(max, solutionSize, solution, result) {
    if (solution.length === solutionSize) {
        result.push(solution.slice());
    } else {
        var current = solution.length > 0 ? solution[solution.length - 1] : 0
        for (var i = current + 1; i <= max; i++) {
            solution.push(i);
            combineHelper(max, solutionSize, solution, result);
            solution.pop();
        }
    }
}
